src/cmd-koji-upload

```diff
@@ -682,33 +683,58 @@ class Upload(_KojiBase):
         """
         serverdir = klib.unique_path(f"{self.build.build_id}-cosa")
         log.debug(f"uploading files to {serverdir}")
+        meta_path = ""
+        meta_remote_path = ""
+        callback = klib._progress_callback
+
         for _, meta in (self.build).get_artifacts():
             local_path = meta['local_path']    # the local file to upload
             remote_path = meta['upload_path']  # the name of the file to upload
             log.info(f"Uploading {local_path} to {serverdir}/{remote_path}")
-            callback = klib._progress_callback
+            if ("meta.json" in local_path) and ("commitmeta.json" not in local_path):
+                # Keep location for meta.json
+                meta_path = local_path
+                meta_remote_path = remote_path
             self.session.uploadWrapper(local_path, serverdir, name=remote_path,
                                        callback=callback)
             if callback:
-                print('')
+                pass
 
         self._uploaded = True
         self._remote_directory = serverdir
+        cginfo = ""
 
-        """ If the final Import fails retry it """
-        for attempt in range(0, self._retry_attempts):
+        # If the final Import fails retry it
+        for attempt in range(1, self._retry_attempts + 1):
+            self.manifest
             try:
+                log.info(f"Finishing Koji Upload - Try {attempt}/{self._retry_attempts}")
                 cginfo = self.session.CGImport(self.manifest, serverdir, token=self._token)
+                log.info(json.dumps(cginfo, sort_keys=True, indent=3))
+                # Koji has some flakes here, confirm the nvr to make sure the upload worked
+                if cginfo['nvr']:
+                    log.info(f"Recorded build {cginfo['nvr']}")
+                    break
             except Exception as e:
-                if attempt < self._retry_attempts - 1:
-                    continue
+                if attempt < self._retry_attempts:
+                    if (self._token):
+                        log.info(f"Koji upload failed with: {e}. Canceling koji reservation \
+                        {self._build_id} and retrying")
+                        self.session.CGRefundBuild("coreos-assembler", self._build_id, self._token, state=4)
+                        log.info(f"Koji id reservation canceled: {self._build_id}")
+                        Reserve(self._profile).reserve_id(self.build)
+                        # Force manifest and checksum/size update for the new meta generated
+                        self._manifest = None
+                        self._image_files = None
+                        self.build._found_files[meta_path]['md5'] = md5sum_file(meta_path)
```

Comment:
I'm not too familiar with the koji code.  What happens if  `meta_path` and `meta_remote_path` are `""`?  I see they are only set `if ("meta.json" in local_path) and ("commitmeta.json" not in local_path)`

Reply:
If the code reaches here, it will never be empty. It was previous set [here]( https://github.com/coreos/coreos-assembler/blob/1183b990f3cdfd223651c49c4bd93fbd26a178e4/src/cmd-koji-upload#L287-L291). I'm rewriting it, since the id reservation changed the meta.json checksum created won't match it anymore for new the meta.json.

Reply:
Got it.  LGTM!